plugins/services/src/js/reducers/serviceForm/VolumeMounts.js

```diff
@@ -57,45 +58,40 @@ module.exports = {
         if (volume.name == null) {
           return memo;
         }
-        mountMap[volume.name] = volumes.push(volume.name) - 1;
-        memo.push(new Transaction(['volumeMounts'], mountMap[volume.name], ADD_ITEM));
-        memo.push(new Transaction([
-          'volumeMounts',
-          mountMap[volume.name],
-          'name'
-        ], volume.name));
-
-        return memo;
+        volumeIndexMap[volume.name] = volumes.push(volume.name) - 1;
+        return memo.concat(
+          new Transaction(['volumeMounts'], volumeIndexMap[volume.name], ADD_ITEM),
+          new Transaction(
+            ['volumeMounts', volumeIndexMap[volume.name], 'name'],
+            volume.name
+          )
+        );
       }, []);
 
       transactions = transactions.concat(volumeTransaction);
     }
 
-    const containerVolumesTransactions =
-      state.containers
-        .reduce(transformContainers, [])
-        .reduce((memo, [containerIndex, volumeMount]) => {
-          const {name, mountPath} = volumeMount;
-
-          if (mountMap[name] == null) {
-            mountMap[name] = volumes.push(name) - 1;
-            memo.push(new Transaction(['volumeMounts'], mountMap[name], ADD_ITEM));
-            memo.push(new Transaction([
-              'volumeMounts',
-              mountMap[name],
-              'name'
-            ], name));
-          }
-
-          memo.push(new Transaction([
-            'volumeMounts',
-            mountMap[name],
-            'mountPath',
-            containerIndex
-          ], mountPath));
+    const containers = state.containers || [];
+    const containerVolumesTransactions = containers
+      .reduce(transformContainers, [])
```

Comment:
Yeah, the nested reduce is not pretty. I feel we can break this into separate functions, but I think we have a few functions like this, that needs a cleanup. I posted a few comments about it here too: https://github.com/dcos/dcos-ui/pull/1524#pullrequestreview-12070726